src/lib/FlightTasks/tasks/AutoLineSmoothVel/FlightTaskAutoLineSmoothVel.cpp

```diff
@@ -246,34 +257,49 @@ void FlightTaskAutoLineSmoothVel::_prepareSetpoints()
 		    PX4_ISFINITE(_position_setpoint(1))) {
 			// Use position setpoints to generate velocity setpoints
 
-			// Get various path specific vectors. */
+			// Get various path specific vectors
 			Vector3f pos_traj;
 			pos_traj(0) = _trajectory[0].getCurrentPosition();
 			pos_traj(1) = _trajectory[1].getCurrentPosition();
 			pos_traj(2) = _trajectory[2].getCurrentPosition();
 			Vector2f pos_traj_to_dest_xy(_position_setpoint - pos_traj);
 			Vector2f u_pos_traj_to_dest_xy(pos_traj_to_dest_xy.unit_or_zero());
-			const bool has_reached_altitude = fabsf(_position_setpoint(2) - pos_traj(2)) < _param_nav_mc_alt_rad.get();
 
-			float speed_sp_track = _getMaxSpeedFromDistance(pos_traj_to_dest_xy.length());
+			// Unconstrained desired velocity vector
+			Vector2f vel_sp_xy = u_pos_traj_to_dest_xy * _mc_cruise_speed;
+
+			Vector2f vel_max_xy;
+			vel_max_xy(0) = _getMaxSpeedFromDistance(fabsf(pos_traj_to_dest_xy(0)));
+			vel_max_xy(1) = _getMaxSpeedFromDistance(fabsf(pos_traj_to_dest_xy(1)));
+
+			const bool has_reached_altitude = fabsf(_position_setpoint(2) - pos_traj(2)) < _param_nav_mc_alt_rad.get();
+			Vector2f vel_min_xy;
 
 			if (has_reached_altitude) {
-				speed_sp_track = math::constrain(speed_sp_track, _getSpeedAtTarget(), _mc_cruise_speed);
+				// Compute the minimum speed in NE frame. This is used
+				// to force the drone to pass the waypoint with a desired speed
+				Vector2f u_prev_to_target_xy((_target - _prev_wp).unit_or_zero());
+				vel_min_xy = u_prev_to_target_xy * _getSpeedAtTarget();
+				vel_min_xy(0) = fabsf(vel_min_xy(0));
+				vel_min_xy(1) = fabsf(vel_min_xy(1));
 
 			} else {
-				speed_sp_track = math::constrain(speed_sp_track, 0.0f, _mc_cruise_speed);
-
+				// The drone has to change altitude, stop at the waypoint
+				vel_min_xy.setAll(0.f);
 			}
 
-			Vector2f vel_sp_xy = u_pos_traj_to_dest_xy * speed_sp_track;
+			// Constrain the norm of each component using min and max values
+			Vector2f vel_sp_constrained_xy;
+			vel_sp_constrained_xy(0) = _maxAbs(_constrainAbs(vel_sp_xy(0), vel_max_xy(0)), vel_min_xy(0));
```

Comment:
```suggestion
			vel_sp_constrained_xy(0) = math::sign(vel_sp_xy(0)) * math::constrain(fabsf(vel_sp_xy(0)), vel_min_xy(0), vel_max_xy(0));
```
min max would need to be abs as well. Just a suggestion.